lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -1358,6 +1358,25 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
     return b.create<math::SqrtOp>(loc, payloadArgs[0]);
   if (isa<AtenRsqrtOp>(op))
     return b.create<math::RsqrtOp>(loc, payloadArgs[0]);
+  if (auto bitwiseAnd = dyn_cast<AtenBitwiseAndTensorOp>(op)) {
+    if (bitwiseAnd.getType()
+            .cast<ValueTensorType>()
+            .getDtype()
+            .isa<mlir::FloatType>()) {
+      bitwiseAnd.emitError("Bitwise_And does not support floating point dtype");
+      return nullptr;
+    }
+    Type selfDtype =
```

Comment:
This constraint is too strong. The bit width of  integer type can be different and it's not hard to add the support. You can use `convertScalarToDtype` to convert the operands to the same type/width first. Refer to
https://github.com/llvm/mlir-npcomp/blob/56c6e3676bbe1571dab0c7cc1fb5080846819438/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1429-L1435

Reply:
I tried converting the operands to be of the same width, but then I got the following error: 
ERROR: value (Tensor with shape=[3, 4] ...) is not close to golden value (Tensor with shape=[3, 4] ...).

This op is doing `Bitwise_And` of each input element of the tensor T1 with that of T2. When we expand the width of a signed integer (say from si32 to si64), due to the integer representation mechanism and sign bit, the output of the `Bitwise_And` operation will not be the same as that of the unexpanded input. I think for this op, we can't do type promotion.

Reply:
Did you use `convertScalarToDtype` to do the conversion? Also it would need the fix in RefineType to work because RefineType need to figure out the correct result type first.  All the operands should be signed extended to the same width and do `bitwise_and` operation.  You can also try with python to verify this `torch.bitwise_and(torch.tensor(3, dtype=torch.int16), torch.tensor(-1, dtype=torch.int8))`
would give you `tensor(3, dtype=torch.int16)`.

Reply:
Hi @cathyzhyi, `convertScalarToDtype` can't be used here, because in this case, both the inputs are tensor. Do we have any other method which does this type of conversion? @ramiro050, any suggestions?

Reply:
> both the inputs are tensor
yea, but `payloadArgs[0]` and `payloadArgs[1]` are scalar elements of each input tensor so `convertScalarToDtype ` can still be used to convert each scalar element to the target data type before doing `arith::AndIOp`. You can refer to https://github.com/llvm/torch-mlir/blob/56c6e3676bbe1571dab0c7cc1fb5080846819438/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1429-L1435.

Reply:
I am getting the following error, even when I do the conversion in the same way as above:
![image](https://user-images.githubusercontent.com/68822896/144046245-f3d9e6bd-6048-4b71-879e-6f690346bf32.png)

Reply:
I have updated the PR with the type conversion changes. Please tell me if you think there's some error in the code. @cathyzhyi @ramiro050